File: codex_token_trace/formatting.py
from __future__ import annotations

from typing import Iterable, Sequence


def fmt_int(value: int | None) -> str:
    if value is None:
        return "-"
    return f"{value:,}"


def fmt_pct(value: float | None) -> str:
    if value is None:
        return "-"
    return f"{value * 100:.1f}%"


def shorten(text: str, width: int) -> str:
    text = text.replace("\n", "\\n")
    if len(text) <= width:
        return text
    return text[: max(0, width - 1)] + "…"


def table(headers: Sequence[str], rows: Iterable[Sequence[object]]) -> str:
    materialized = [[str(cell) for cell in row] for row in rows]
    widths = [len(h) for h in headers]
    for row in materialized:
        for i, cell in enumerate(row):
            widths[i] = max(widths[i], len(cell))

    def render(row: Sequence[str]) -> str:
        return "  ".join(cell.ljust(widths[i]) for i, cell in enumerate(row))

    lines = [render(headers), render(["-" * w for w in widths])]
    lines.extend(render(row) for row in materialized)
    return "\n".join(lines)

File: src/formatting.ts
export const formatInt = (value: number | null | undefined): string => {
  if (value === null || value === undefined) {
    return "-";
  }
  return new Intl.NumberFormat("en-US").format(value);
};

export const formatPct = (value: number | null | undefined): string => {
  if (value === null || value === undefined) {
    return "-";
  }
  return `${(value * 100).toFixed(1)}%`;
};

export const shorten = (value: string, width: number): string => {
  const text = value.replaceAll("\n", "\\n");
  if (text.length <= width) {
    return text;
  }
  return `${text.slice(0, Math.max(0, width - 1))}…`;
};

export const table = (headers: string[], rows: Array<Array<string | number>>): string => {
  const stringRows = rows.map((row) => row.map(String));
  const widths = headers.map((header, index) =>
    Math.max(header.length, ...stringRows.map((row) => row[index]?.length ?? 0)),
  );

  const render = (row: string[]): string =>
    row.map((cell, index) => cell.padEnd(widths[index] ?? 0)).join("  ");

  return [
    render(headers),
    render(widths.map((width) => "-".repeat(width))),
    ...stringRows.map(render),
  ].join("\n");
};
